### DISCUSSION TITLE
A few questions about the object serialization example

### DISCUSSION THREAD

**Discussion by jpoesen** [mglaman/drupal-typed-data-by-example#16] [Q&A]
Thanks for the serialization example (`serialization/object.php`), Matt.

A few questions, if I may:

## 1. Service instantiation
Why: 
```php
$serializer = \Drupal::getContainer()->get('serializer');
```

And not:
```php
$serializer = \Drupal::service('serializer');
```

## 2. Assert statements
`Assert` statements aren't very common in Drupal code. I've only seen them in (your?) drupal commerce code.  

Would you use all of the example's `assert` statements in production code, or are you using them liberally here to show how you can use them if you want to?

```php
$serializer = \Drupal::getContainer()->get('serializer');
assert($serializer instanceof Serializer);
```
--> this would most likely only fail if the serializer module is not enabled?

## 3. 'Should I use this' question
Say you want to export a subset of a content entity's fields into a specific JSON structure, different from what default serialization would produce. 

Would you manually build the array or use a data definition map like you did in the example?

Since the entity system already uses typed data, I assume there's no need to re-specify each field/array member's data type? Maybe I'm missing or not correctly understanding something. Apologies if this question doesn't make sense.

Thanks!

**Comment by mglaman** [comment]
## 1. Service instantiation

Because the PhpStorm plugins for Symfony & Symfony Drupal Bridge get the return type messed up from the service container with `\Drupal::service`. So it's a habit I've picked up over the past few years. All because `::service` has `@return mixed`, which breaks the parent return.

## 2. Assert statements

I think `/** @var */` type hints are ugly and they don't enforce actual expectations. If an assertion fails, so does the script. This forces static analysis to know what the value actually is (for PhpStorm and PHPStan.)

I don't know if it's the best way to communicate in the examples. I wasn't sure, so I went with this for now. Sometimes I see things as PHPUnit tests with `assertEquals` but this seemed less convoluted. I am totally open to feedback on this.

## 3. 'Should I use this' question

This needs to be added to the file. When working with Typed Data at this level, you usually have abstract data that isn't in the entity API. The fact I used a user profile probably didn't help. The best example is what I wrote for Simplytest and what I need to add here.

We have a Launch endpoint with a specific payload: 

```json
{
  "project": {...project schema},
  "drupalVersion: "",
  "installProfile": "",
  "manualInstall": "",
  "additionalProjects": [...array of project schema]
}
```

They're here:

* https://git.drupalcode.org/project/simplytest/-/blob/8.x-4.x/modules/simplytest_launch/src/TypedData/InstanceLaunchDefinition.php
* https://git.drupalcode.org/project/simplytest/-/blob/8.x-4.x/modules/simplytest_launch/src/TypedData/ProjectInfoDefinition.php

Our controller takes the `POST` JSON and create a data type object for us to perform validation of the request payload

```php
  /**
   * Project launcher service for react.
   */
  public function launchProject(Request $request) {
    $content = $request->getContent();
    $submission = Json::decode($content);
    $this->validateSubmission($submission);
```

I had to wrap in `validateSubmission` to handle some legacy logic not yet in constraints. But this is the main part

```php
  private function validateSubmission($data) {
    // @todo Flood protection preflight check (IP based, possibly a problem.)
    try {
      $definition = $this->typeDataManager->create(InstanceLaunchDefinition::create(), $data);
    }
    catch (\Throwable $e) {
      throw new ServiceUnavailableHttpException(null, $e->getMessage());
    }
    $constraints = $definition->validate();
    if ($constraints->count() > 0) {
      $exception = new UnprocessableHttpEntityException();
      $exception->setViolations($constraints);
      throw $exception;
    }
```

If the data isn't valid per our constraints, we reject the request. Our controller merely decodes the request content and has the Typed Data API validate it.
